plugins/net_plugin/include/eos/net_plugin/message_buffer.hpp

```diff
@@ -141,6 +141,9 @@ namespace eosio {
      */
     void advance_write_ptr(uint32_t bytes) {
       advance_index(write_ind, bytes);
+      while (write_ind.first >= buffers.size()) {
+        buffers.push_back(pool().malloc());
```

Comment:
This seems "too late".

Our usage pattern is to pass the boost buffers to `async_read_some` and then in the resulting handler, call `advance_write_ptr` to account for the data that was written in the async call.  

By this time, the data that would have been written has already over flowed any buffer it thought it had.

Reply:
I agree as an API this makes sense to advance, then write.  So this change is good outright but not exercised in our code base

Reply:
There will not be any overflow as async_read_some() will stop when the allocated buffers are full. The problem here was that you could end up in a situation where write_ind pointed past the end of buffers which caused the problems when generating the boost mutable buffers.

Reply:
> you could end up in a situation where write_ind pointed past the end of buffers

have we identified this situation though?  The net-plugin never "directly writes" it only indirectly writes through these boost buffers and then returns how much it had written. In turn this is the only call to `advance_write_ptr`.  So, either the calling code is not respecting the conventions of boost and boost was giving us a larger bytes read than it actually wrote to, or there is some other problem going on.

Reply:
Let me be clearer, i wanted to maintain the invariant that write_ptr() always points to allocated memory. This change assures that.

Reply:
If that is the case, then I suggest that `advance_write_ptr` which is used to record a write _that has already happened_  ASSERT that it is not advancing past the end.   Its invariant is that you aren't trying to advance passed your requested space.

`write_ptr()` should be the method that maintains about what it can/cannot return.  So, detect a write pointer at the end of the buffer and expand it there. 

This way, 
 * you don't waste a segment of memory just to store a "Full" buffer. 
 * you notify the callers when they are doing something the shouldn't be doing (like advancing a write pointer past where they have requested space via `add_space` or `add_buffer_to_chain`)
 * and you maintain the invariant you are worried about

In all honesty, we already have 2 methods for explicitly requesting space and doing it implicitly on either `write_ptr()` or `advance_write_ptr()`  just seems unnecessary.  That we are adding this to either as opposed to throwing asserts for mis-use of the explicit API is just going to cover up caller errors creating a tight coupling between the callers and some undocumented implicit behavior.

Reply:
I added a unit test that exposed the problem (and would cause your assert when generating boost buffers). If boost async was filling the current buffers (as it would in a stress test) and we did not read all the data, we would pass invalid pointers on the next async_read_some() call.

Reply:
I am not interested in delaying the fix for the test net any longer.

Reply:
I could add the buffer growth in two other places: write_ptr() and get_buffer_sequence_for_boost_async_read() which are the two places that the pointer indicated by write_ind can escape the class, but doing it in advance_write_ptr() seemed to be more consistent. If you have read enough to fill your buffer, you're probably going to need another buffer in the sequence to contain the rest of a partial message (unless you were lucky and got a whole message). So in the nominal case it won't be an extra buffer allocation.